Integration of PVeRA

leofillioux opened the pull request:
This PR is a continuation of issue #2948, for which we proposed the integration of the [PVeRA adapter](https://www.arxiv.org/abs/2512.07703).

As recommended in the issue, we based our implementation on the implementation of the VeRA adapter, as both adapters are very close. Here are a list of the contributions from this PR.
- Added the PVeRA adapter folder in src/peft, and adapted the `config.py`, `layer.py`, and `model.py` files from the ones from VeRA.
- Added some documentation.
- Adapted the `tests/test_vera.py` to `tests/test_pvera.py` and made sure this ran properly.

Note: because I'm running on a Mac, I was not able to run `make test` (had an error with MPS).

@BenjaminBossan could you please give me some feedback?

leofillioux committed:
Integration of PVeRA. Added the implementation of PVeRA, the documentation, and the tests.

githubnemo reviewed (commented):
Thanks for the PR, this already looks quite good!

Please check the copyright notices and make sure that it is up-to-date (it often says 2024 but it should be 2025).

I've done a quick review and left a few comments. General remarks:

- let's add PVeRA to `tests/test_custom_models.py` (adding the important configurations to `TEST_CASES`, similar to VeRA)
- if these pass we can extend the coverage by adding PVeRA to `tests/test_decoder_models.py` and `tests/test_encoder_decoder_models.py`

Once these are implemented I'll do a more thorough review. After that it'd be nice to have a runnable example and to integrate it into the method comparison to benchmark it against the other methods (and to check our expectations).

Heads up: I'll be rather off than on in the next days, so merging and review will most likely happen in the next year.

leofillioux committed:
Recommendations from the PR review.

leofillioux committed:
Added PVeRA tests to custom_models, custom_decoder, and custom_encoder_decoder.

githubnemo commented on src/peft/tuners/pvera/config.py:
```diff
@@ -0,0 +1,172 @@
+    """
+    This is the configuration class to store the configuration of a [`PVeRAModel`].
+
+    Paper: https://www.arxiv.org/abs/2512.07703.
```
```suggestion
    Paper: https://huggingface.co/papers/2512.07703.
```

githubnemo commented on src/peft/tuners/pvera/config.py:
```diff
@@ -0,0 +1,172 @@
+    Args:
+        r (`int`, *optional*, defaults to `256`):
+            PVeRA parameter dimension ("rank"). Choose higher values than LoRA ranks here, since PVeRA uses far fewer
+            parameters than LoRA.
```
```suggestion
            PVeRA parameter dimension ("rank"). Choose higher values than LoRA ranks here, since PVeRA uses parameter sharing and therefore uses far fewer
            parameters than LoRA.
```

or something like that to explain where the reduced parameter usage comes from.

githubnemo reviewed (commented):
Hey, thanks for the update!

Implementation and tests look quite mature. I think that if we provide bitsandbytes support we also need at least one test in `tests/test_gpu_examples.py` for quality control. I think `test_causal_lm_training_4bit_vera` could be used as a base.

In general I think we should rename `PVeRA*` to `Pvera*` (e.g., `PVeRAModel` -> `PveraModel`) to be consistent with `VeraModel` and friends. It is quite hard to remember the spelling of the various abbreviations already :)

Rest of the review is in the comments.

After the comments are resolved and the CI is green I think it would be nice to integrate `PVeRA` into the MetaMathQA benchmark by adding an experiment file based on the [VeRA experiment](https://github.com/huggingface/peft/tree/main/method_comparison/MetaMathQA/experiments/vera/llama-3.2-3B-default).

Edit: To fix the docs build, add the PVeRA entry to `docs/source/_toctree.yml` similar to the other entries.

leofillioux committed:
Added recommendations from PR.

leofillioux commented on tests/testing_common.py:
```diff
@@ -1114,7 +1117,7 @@ def _test_training_layer_indexing(self, model_id, config_cls, config_kwargs):
             model = self.transformers_class.from_pretrained(model_id)
             model = get_peft_model(model, config)
-            model = model.to(self.torch_device)
+            model = model.to(self.torch_device).eval()
```
Yes you're right, it doesn't make sense to put `.eval()`. However there is a problem in that `model` is in training mode here, whereas `model_from_pretrained` is is eval mode. This means that for PVeRA one model will sample from the learned distribution and the other will not, therefore giving different results, which is why I had put them both in eval mode. Do you think we should maybe just remove this test for PVeRA?

githubnemo commented on src/peft/tuners/pvera/__init__.py:
```diff
@@ -0,0 +1,26 @@
+from .model import PveraModel
+
+
+__all__ = ["Linear", "PVeRALayer", "PveraConfig", "PveraModel"]
```
```suggestion
__all__ = ["Linear", "PveraLayer", "PveraConfig", "PveraModel"]
```

+for the class itself as well

githubnemo commented on tests/test_custom_models.py:
```diff
@@ -2113,6 +2141,8 @@ def test_parameters_after_loading_model(self, test_name, model_id, config_cls, c
         elif issubclass(config_cls, (VBLoRAConfig, RandLoraConfig, OSFConfig)):
             lr = 0.01  # otherwise we get nan
+        elif issubclass(config_cls, PVeRAConfig):  # needs very small lr to not get nan
+            lr = 1e-6
```
OK. Let's extend the comment to be more precise:

```
# needs a very small lr to not get nan in pvera_lambda_b due to high input values in this test (up to 90)
```

githubnemo reviewed (commented):
Thanks for the changes!

I've commented on hopefully all the outstanding issues and added a few nits.

CI seems to be passing except for the `.eval` issue which is hopefully resolved with the proposed fix.

leofillioux committed:
Suggestions from PR (still need to analyze redundent tests and change sample_at_inference).

leofillioux committed:
Suggestions from PR (still need to analyze redundent tests and change sample_at_inference).

leofillioux committed:
Changed the sample_at_inference variable to allow dict.

leofillioux commented on tests/test_pvera.py:
```diff
@@ -0,0 +1,302 @@
+        output_other_loaded = peft_model(input)
+
+        assert torch.allclose(output_default, output_default_loaded, atol=1e-3, rtol=1e-3)
+        assert torch.allclose(output_other, output_other_loaded, atol=1e-3, rtol=1e-3)
```
To be honest, I integrated these (`test_multiple_adapters_save_load_save_projection_true` and `test_multiple_adapters_save_projection_true_contains_pvera_A_pvera_B`) because they were integrated for VeRA, but I agree that they do seem a bit redundant. I'm ok with removing them if you think they aren't useful.

githubnemo replied:
Yeah, let's remove them.

githubnemo reviewed (commented):
Thanks for the updates!

Correct me if I'm wrong but I think there's currently no test that checks whether the output is non-deterministic during training and during inference when requested. Let's make sure that this code path is tested, at least roughly.

I checked the old comments and resolved those that are done. 
There are still some open comments:
- https://github.com/huggingface/peft/pull/2952#discussion_r2712674190
- https://github.com/huggingface/peft/pull/2952/files#r2686810130.

Once that is done, let's add an experiment for the MetaMathQA method comparison. You can look at `method_comparison/MetaMathQA/experiments/vera/llama-3.2-3B-default/` for inspiration. You  may need to tune some hyper-parameters since they are probably non-optimal. Note that we try to keep the number of parameters roughly equal between methods. There's also documentation there how to run these experiments yourself for quick iteration.

It'd also good to have an example, maybe showcasing how to leverage sampling for getting confidence intervals? I let you be the judge of what is most effective in show-casing PVeRA.

leofillioux committed:
Remove test_multiple_adapters_save_load_save_projection_true and test_multiple_adapters_save_projection_true_contains_*_A_*_B tests.

leofillioux committed:
Use docstring values for the help values.

leofillioux commented on docs/source/package_reference/pvera.md:
```diff
@@ -0,0 +1,40 @@
+
+# PVeRA: Probabilistic Vector-Based Random Matrix Adaptation
+
+[PVeRA](https://huggingface.co/papers/2512.07703) is a parameter-efficient fine-tuning technique that is base on VeRA, in the family of the LoRA-based adapters. It keeps the very low parameter budget of VeRA, but increases the performance by learning a distribution of latent adaptations. This also enables models adapted with PVeRA to generate Monte Carlo confidence interval estimates, by sampling ...
```
This is done in the docstring from PveraModel, so it should appear in the documentation.

leofillioux committed:
Added default experiment parameters for PVeRA.

leofillioux committed:
Changed the lr for the MetaMathQA, and ran the expriment.

leofillioux committed:
Added an example of confidence interval generation using the probabilistic nature of PVeRA.

githubnemo reviewed (commented):
Hey @leofillioux!

I think this is almost good to go.

Thanks for providing the benchmark experiment + results.
The results are in line with the VeRA (~36% accuracy). If you already have a bit of experience in hyper-parameter tuning VeRA/PVeRA (i.e. maybe you know that the learning rate usually needs to be a bit higher) you're welcome to integrate that knowledge into the benchmark but this doesn't need to happen in this PR - it's just a hint that the VeRA configuration might not be optimal.

There's still the open issue of per-adapter `sample_at_inference` setting (via https://github.com/huggingface/peft/pull/2952#discussion_r2686785742). I think it is not that big of a deal to not have adapter-wise sampling, we can leave it out.

The failing CI is unrelated but make sure to merge with main so that we can give the CI a final run after you've addressed the few remaining comments.

leofillioux committed:
Updates from code review.

leofillioux committed:
Merge branch 'huggingface:main' into main

leofillioux commented:
Hello @githubnemo,

Thanks again for the review! I've addressed the remaining points. Regarding the `sample_at_inference` as a dictionary, this is normally handled in [this commit](https://github.com/huggingface/peft/pull/2952/changes/aaa267f9f0d2a0ebb0738215660030539e4d52a4#diff-426fd3851b6a426c7e10876755f446f883852624f04dc409206b32aee298b3ee). I've also merged with main.

githubnemo reviewed (approved):
> Hello @githubnemo,
> 
> Thanks again for the review! I've addressed the remaining points. Regarding the `sample_at_inference` as a dictionary, this is normally handled in [this commit](https://github.com/huggingface/peft/pull/2952/changes/aaa267f9f0d2a0ebb0738215660030539e4d52a4#diff-426fd3851b6a426c7e10876755f446f883852624f04dc409206b32aee298b3ee). I've also merged with main.

Thanks for the quick response. I think I was a bit vague with what I meant regarding `sample_at_inference` - possibly because we have two definitions of adapter. I meant 'adapter' in the sense of PEFT's `.add_adapter()`, i.e. a second `PVeraConfig` on the same model. In that case `sample_at_inference` would be a global setting and I couldn't disable it for adapter 1 without disabling it for adapter 2 as well. I was not talking about individual adapted layers.

But I think this is fine as is! I'll let the CI run, if it passes I'll merge this if you don't have any objections :)

BenjaminBossan commented:
@leofillioux Some of the PVeRA GPU tests are failing on CI. This can be tested like so:

```sh
# single gpu
pytest -m single_gpu_tests tests/test_gpu_examples.py -k pvera -v
# multi gpu
pytest -m multi_gpu_tests tests/test_gpu_examples.py -k pvera -v
```

One error is trivial (using `vera_dropout` instead of `pvera_dropout`) but for the others, I'm not so sure, as we get this error in the forward pass:

`RuntimeError: mat1 and mat2 shapes cannot be multiplied (128x32 and 16x4096`

Do you have time to investigate this?

leofillioux commented:
Hello @BenjaminBossan, I tried to reproduce this error but I only get this error

```
ValueError: Asking to pad but the tokenizer does not have a padding token. Please select a token to use as `pad_token` `(tokenizer.pad_token = tokenizer.eos_token e.g.)` or add a new pad token via `tokenizer.add_special_tokens({'pad_token': '[PAD]'})`.
```

I also get the same error when running `-k vera`. I'm guessing this comes from my environment, but it fails even after reinstalling from scratch. Any idea how to fix this?

BenjaminBossan commented:
Thanks for checking @leofillioux. What package versions are you using: I'm using

- PEFT main
- transformers main / v5.5.4
- bitsandbytes 0.49.2
- torch 2.10.0

BenjaminBossan commented:
@leofillioux Did you have time to check yet?

leofillioux commented:
@BenjaminBossan I wasn't able to reproduce the error. I haven't had much time in the last few days, but I'll have more time in the end of the week!

BenjaminBossan commented:
Thanks. In the meantime, if you could share the package versions of your env, I could also do more testing.

leofillioux commented:
Yes sure, here is the full output of `pip freeze`.
```
accelerate==1.13.0
aiohappyeyeballs==2.6.1
aiohttp==3.13.5
aiosignal==1.4.0
annotated-doc==0.0.4
anyio==4.13.0
async-timeout==5.0.1
attrs==26.1.0
bitsandbytes==0.49.2
black==26.3.1
certifi==2026.2.25
charset-normalizer==3.4.7
click==8.3.2
coverage==7.13.5
cuda-bindings==12.9.4
cuda-pathfinder==1.5.2
cuda-toolkit==13.0.2
datasets==4.8.4
diffusers==0.37.1
dill==0.4.1
exceptiongroup==1.3.1
execnet==2.1.2
fastjsonschema==2.21.2
filelock==3.25.2
frozenlist==1.8.0
fsspec==2026.2.0
gitdb==4.0.12
GitPython==3.1.46
h11==0.16.0
hf-doc-builder==0.5.0
hf-xet==1.4.3
httpcore==1.0.9
httpx==0.28.1
huggingface_hub==1.10.1
idna==3.11
importlib_metadata==9.0.0
iniconfig==2.3.0
Jinja2==3.1.6
joblib==1.5.3
jsonschema==4.26.0
jsonschema-specifications==2025.9.1
jupyter_core==5.9.1
markdown-it-py==4.0.0
MarkupSafe==3.0.3
mdurl==0.1.2
mpmath==1.3.0
multidict==6.7.1
multiprocess==0.70.19
mypy_extensions==1.1.0
nbformat==5.10.4
networkx==3.4.2
numpy==2.2.6
nvidia-cublas==13.1.0.3
nvidia-cublas-cu12==12.8.4.1
nvidia-cuda-cupti==13.0.85
nvidia-cuda-cupti-cu12==12.8.90
nvidia-cuda-nvrtc==13.0.88
nvidia-cuda-nvrtc-cu12==12.8.93
nvidia-cuda-runtime==13.0.96
nvidia-cuda-runtime-cu12==12.8.90
nvidia-cudnn-cu12==9.10.2.21
nvidia-cudnn-cu13==9.19.0.56
nvidia-cufft==12.0.0.61
nvidia-cufft-cu12==11.3.3.83
nvidia-cufile==1.15.1.6
nvidia-cufile-cu12==1.13.1.3
nvidia-curand==10.4.0.35
nvidia-curand-cu12==10.3.9.90
nvidia-cusolver==12.0.4.66
nvidia-cusolver-cu12==11.7.3.90
nvidia-cusparse==12.6.3.3
nvidia-cusparse-cu12==12.5.8.93
nvidia-cusparselt-cu12==0.7.1
nvidia-cusparselt-cu13==0.8.0
nvidia-nccl-cu12==2.27.5
nvidia-nccl-cu13==2.28.9
nvidia-nvjitlink==13.0.88
nvidia-nvjitlink-cu12==12.8.93
nvidia-nvshmem-cu12==3.4.5
nvidia-nvshmem-cu13==3.4.5
nvidia-nvtx==13.0.85
nvidia-nvtx-cu12==12.8.90
packaging @ file:///home/conda/feedstock_root/build_artifacts/bld/rattler-build_packaging_1769093650/work
pandas==2.3.3
parameterized==0.9.0
pathspec==1.0.4
-e git+https://github.com/huggingface/peft.git@efe0fe6acd72cb3bf1ebfc807c159bf0b9481f5e#egg=peft
pillow==12.2.0
platformdirs==4.9.6
pluggy==1.6.0
propcache==0.4.1
protobuf==7.34.1
psutil==7.2.2
pyarrow==23.0.1
Pygments==2.20.0
pytest==9.0.3
pytest-cov==7.1.0
pytest-xdist==3.8.0
python-dateutil==2.9.0.post0
pytokens==0.4.1
pytz==2026.1.post1
PyYAML==6.0.3
referencing==0.37.0
regex==2026.4.4
requests==2.33.1
rich==15.0.0
rpds-py==0.30.0
ruff==0.12.12
safetensors==0.7.0
scikit-learn==1.7.2
scipy==1.15.3
sentencepiece==0.2.1
shellingham==1.5.4
six==1.17.0
smmap==5.0.3
sympy==1.14.0
threadpoolctl==3.6.0
tokenizers==0.22.2
tomli==2.4.1
torch==2.10.0
torchvision==0.26.0
tqdm==4.67.3
traitlets==5.14.3
transformers==5.5.4
triton==3.6.0
typer==0.24.1
typing_extensions==4.15.0
tzdata==2026.1
urllib3==2.6.3
xxhash==3.6.0
yarl==1.23.0
zipp==3.23.0
```

BenjaminBossan commented:
Thanks for sharing. This is strange, as all the important library versions match with what I have in my env (I also tried this exact PEFT commit that you use). Could you please try the following:

1. Add a breakpoint [here](https://github.com/huggingface/peft/blob/76c37d4686cf7245a9a6998fd1e2619d62c9322d/src/peft/tuners/pvera/bnb.py#L396)
2. Run `pytest -m single_gpu_tests tests/test_gpu_examples.py -k pvera`
3. Check the shapes of the relevant tensors

I get:

```python
(Pdb) pvera_A.shape
torch.Size([32, 4096])
(Pdb) pvera_B.shape
torch.Size([4096, 16])
(Pdb) x_temp.shape
torch.Size([4, 32, 4096])
(Pdb) sliced_A.shape
torch.Size([32, 4096])
(Pdb) torch.nn.functional.linear(x_temp, sliced_A).shape
torch.Size([4, 32, 32])
(Pdb) sliced_B.shape
torch.Size([4096, 16])
```

which is clearly incompatible. What do you see?

leofillioux commented:
I'm not able to even get to the breakpoint, I get this error.

```
ModuleNotFoundError: Could not import module 'Seq2SeqTrainer'. Are this object's requirements defined correctly?
```

However, given the shapes, I think I see what is going on. Basically for PVeRA, we have $$\mathbf A \in \mathbb R^{2r, d}$$ (compared to $$\mathbf A \in \mathbb R^{r, d}$$ for VeRA). During training, this is used to generate $$\boldsymbol \mu, \boldsymbol \sigma \in \mathbb R^{r, d}$$, which define the learned distribution parameters, we then sample $$\mathbf z \sim \mathcal N(\boldsymbol \mu, \boldsymbol \sigma^2) \in \mathbb R^{r,d}$$. During inference, we don't sample, and we take $$\mathbf z = \boldsymbol \mu$$. When we want to merge the weights, we only take the first part of $$\mathbf A$$, which is responsible for generating $$\boldsymbol \mu$$. This is actually done [here](https://github.com/huggingface/peft/blob/76c37d4686cf7245a9a6998fd1e2619d62c9322d/src/peft/tuners/pvera/layer.py#L251). I think that just this part is missing for this file.

I hope that my explanation is clear. I'll try to fix my environment, but in the meantime, I think this should fix this issue. Were there any others (other than the wrong name for `pvera_dropout`)?

BenjaminBossan commented:
That could be it. The code you linked is only used during the merging/unmerging path but not the forward step. Would the correct change then be something like:

```python
sliced_A = pvera_A[: self.r[active_adapter], : self.in_features].to(x.device)
sliced_lambda_d = lambda_d[:self.r[active_adapter]]
x_temp = dropout(x.to(lambda_d.dtype))
adapter_output = lambda_b * torch.nn.functional.linear(
    sliced_lambda_d * torch.nn.functional.linear(x_temp, sliced_A), sliced_B
)
```

But that would lead to different results, would it not?

leofillioux commented:
Ah yes, sorry! Then the problem is that I try to apply the adapter directly like VeRA, and don't perform the sampling, which is why there is a mismatch in the dimensions. I think the correct change would be something like this.

```python
x_temp = dropout(x.to(lambda_d.dtype))

mu, logvar = (lambda_d * F.linear(dropout(x_temp), sliced_A)).chunk(2, dim=-1)
adapter_output = lambda_b * F.linear(
   self._reparametrize(mu, logvar, self.sample_at_inference), sliced_B
)
```
